src/core/localize.rkt

```diff
@@ -5,89 +5,84 @@
-
-  errs)
-
 ;; Returns a list of expressions sorted by increasing local error
 (define (localize-error prog ctx)
-  (define errs (compute-local-errors-fast prog ctx))
+  (define errs (compute-local-errors prog ctx))
   (sort
     (reap [sow]
           (for ([(expr err) (in-hash errs)])
             (unless (andmap (curry = 1) err)
               (sow (cons err expr)))))
     > #:key (compose errors-score car)))
 
+(define (all-subexpressions expr repr)
+  (remove-duplicates
+  (reap [sow]
+        (let loop ([expr expr] [repr repr])
+          (sow (cons expr repr))
+          (match expr
+            [(? number?) (void)]
+            [(? variable?) (void)]
+            [`(if ,c ,t ,f)
+              (loop c (get-representation 'bool))
+              (loop t repr)
+              (loop f repr)]
+            [(list op args ...)
+              (define atypes (operator-info op 'itype))
+              (for ([arg args] [atype atypes])
+                (loop arg atype))])))))
+
+
 ; Compute local error or each sampled point at each node in `prog`.
 (define (compute-local-errors prog ctx)
-  (define expr (program-body prog))
-  (define subexprs (all-subexpressions expr))
-  (define subexprs-fn
-    (for/hash ([expr (in-list subexprs)])
-      (define subexpr-prog `(λ ,(program-variables prog) ,expr))
-      (define ctx* (struct-copy context ctx [repr (repr-of expr ctx)]))
-      (values expr (eval-prog-real subexpr-prog ctx*))))
-  (define errs (make-hash (map (curryr cons '()) subexprs)))
-  (for ([(pt ex) (in-pcontext (*pcontext*))])
-    (define exacts-hash
-      (for/hash ([expr (in-list subexprs)])
-        (define fn (hash-ref subexprs-fn expr))
-        (values expr (apply fn pt))))
-    (for ([expr (in-list subexprs)])
-      (define err
-        (match expr
-          [(? number?) 1]
-          [(? variable?) 1]
-          [`(if ,c ,ift ,iff) 1]
-          [(list f args ...)
-           (define repr (operator-info f 'otype))
-           (define argapprox
-             (for/list ([arg (in-list args)] [repr (in-list (operator-info f 'itype))])
-               (hash-ref exacts-hash arg)))
-           (ulp-difference (hash-ref exacts-hash expr)
-                           (apply (operator-info f 'fl) argapprox) repr)]))
-      (hash-update! errs expr (curry cons err))))
+  (define expression (program-body prog))
+  (define sub-expressions (all-subexpressions expression (context-repr ctx)))
+  (define prog-list 
+    (for/list ([subexpr (in-list sub-expressions)])
+      `(λ ,(context-vars ctx) ,(car subexpr))))
+  (define ctx-list 
+    (for/list ([subexpr (in-list sub-expressions)])
+      (struct-copy context ctx [repr (cdr subexpr)])))
+
+  (define subexprs-fn (eval-prog-list-real prog-list ctx-list))
 
-  errs)
+  ; Mutable error hack, this is bad
+  (define temp-errs
+    (for/hash ([subexpr (in-list sub-expressions)])
+      (values (car subexpr) '())))
+  (define errs (make-hash))
+  (for ([(k v) temp-errs]) 
+    (hash-set! errs k v))
```

Comment:
Let's just do something like this:
```
(define temp-errs
  (make-hash
    (for/list ([subexpr (in-list sub-expressions)])
      (cons (car subexpr) '()))))
```

Note the use of `(cons foo '())`, because in this case it's a key/value pair.